zerver/views/message_fetch.py

```diff
@@ -808,13 +824,37 @@ def get_messages_backend(request: HttpRequest, user_profile: UserProfile,
             MAX_MESSAGES_PER_FETCH,
         ))
 
-    if user_profile.realm.email_address_visibility != Realm.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
+    # Web public queries are only allowed for limited types of narrows.
+    web_public_filter = {'operator': 'streams', 'operand': 'web-public', 'negated': False}
+    is_web_public_query = False
+    if narrow is not None and web_public_filter in narrow:
+        if not is_web_public_compatible(narrow):
+            return json_unauthorized()
+
+        is_web_public_query = True
+    else:
+        if not request.user.is_authenticated:
+            return json_unauthorized()
+
+    user_realm = None
+    if is_web_public_query:
+        user_realm = get_realm_from_request(request)
+    else:
+        user_realm = user_profile.realm
+
+    if is_web_public_query or \
+            user_profile.realm.email_address_visibility != Realm.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
```

Comment:
I suppose we'll also want even with `EMAIL_ADDRESS_VISIBILITY_EVERYONE`  to ensure the `display_recipient` objects we return don't have real email addresses here.  May end up being an annoying refactoring to make that happen correctly; though I suppose we could for now just not allow web public streams with EMAIL_ADDRESS_VISIBILITY_EVERYONE?

We might want to proceed with creating an organization-level setting for who can create web-public streams, with options being the usual nobody/owners/admins/full members/members before pushing much further than this; we'll want the default configuration to not allow their creation so corporate organizations don't need to think about issues around this featureset.

@sahil839 would you up for doing that in a separate PR?

Reply:
Yeah sure.